### PR TITLE
Upgrade AAVE strategy to AAVE v2

### PR DESCRIPTION
**Pull request by DanielVF**
Upgrade AAVE strategy to use AAVE v2.

[Discussion for upgrade](https://github.com/OriginProtocol/origin-dollar/issues/496)

We should plan on rolling this out as a new strategy address, just to guarantee a clean slate and that all addresses/state are correct. We'll remove the old strategy, then add this strategy.

### Progress:

- [x] Interface changes
- [x] Mock changes
- [x] Contract changes
- [x] Tests pass
- [x] Deployment written
- [x] Deployment tested on fork with fork funds

### Contract change checklist:
  - [x] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [x] Unit tests pass
  - [x] Slither tests pass with no warning
  - [x] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

**Comment by DanielVF** [comment]
Will write deploy script once we have the new deploy library running and merged.

**Comment by DanielVF** [comment]
## Notes on AAVE Liquidity rewards:

- Official [Liquidity Mining Docs](https://docs.aave.com/developers/guides/liquidity-mining#setclaimer)
- Official [Staking Docs](https://docs.aave.com/developers/protocol-governance/staking-aave)
- rewards contract: 0xd784927Ff2f95ba542BfC824c8a8a98F3495f6b5
- stkAAVE: 0x4da27a545c0c5b758a6ba100e3a049001de870f5

### 1. Get rewards amount

`rewards.getRewardsBalance(address[] calldata assets, address user)`

### 2. Convert rewards into stkAAVE

`rewards.claimRewards(address[] calldata assets, uint256 amount, address to)`

### 3. Trigger cooldown

stkAAVE cannot be turned into AAVE without a five day cool down. Sending more funds in restarts the cool down, as does claiming rewards. Additionally, two days after the cool down unlocks, it locks again automatically. To restart the cool down clock:

`stkAAVE.cooldown()`


To find out your current cool down start time:

`stkAAVE.cooldown.stakersCooldowns(address(this))`

### 4. Convert to AAVE

When the time is right:

```
stkAave.claimRewards(address(this), type(uint256).max) 
```

```
stkAave.redeem(address(this), stkAaveBalance)
```

### All together

Because moving funds or claiming rewards resets the cool down, the order matters a little bit

1. First we convert stkAave.redeem to AAVE
2. Then we claim rewards from stkAave
3. Transfer AAVE to the vault to be sold
4. start the cooldown for the next round.

**Review by franckc — APPROVED** [review]
LGTM

**Comment by DanielVF** [comment]
## Testing out AAVE Staking


```javascript
stkAave = Contract.from_explorer("0x4da27a545c0c5b758a6ba100e3a049001de870f5")
aave = Contract.from_explorer("0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9")
incentivesController = Contract.from_explorer("0xd784927Ff2f95ba542BfC824c8a8a98F3495f6b5")
VAULT = "0xE75D77B1865Ae93c7eaa3040B038D7aA7BC02F70"
ADAI_V2 = '0x028171bca77440897b824ca71d1c56cac55b68a3'
ME = "0x4a49985b14bd0ce42c25efde5d8c379a48ab02f3"
FROM_ME = {'from': ME}


stkAave.cooldown(FROM_ME)
chain.sleep(stkAave.COOLDOWN_SECONDS()+100)
chain.mine()

cooldown = stkAave.stakersCooldowns(ME)
windowStart = cooldown + stkAave.COOLDOWN_SECONDS()
windowEnd = windowStart + stkAave.UNSTAKE_WINDOW()
currentTimestamp = chain.time()
currentTimestamp > windowStart and currentTimestamp < windowEnd

stkAaveBalance = stkAave.balanceOf(ME)
rewardLiquidationThreshold = 50
stkAaveBalance > rewardLiquidationThreshold

stkAave.redeem(ME, stkAaveBalance, FROM_ME)
stkAave.balanceOf(ME)

aaveBalance = aave.balanceOf(ME)
aave.transfer(VAULT, aaveBalance, FROM_ME)

pendingRewards = incentivesController.getRewardsBalance([ADAI_V2], ME)
incentivesController.claimRewards([ADAI_V2], pendingRewards, ME, FROM_ME)
```

**Comment by DanielVF** [comment]
Now we have tests, next up is deploy scripts.

**Comment by DanielVF** [comment]
Code for testing rewards

```python
from world import *
from addresses import *
vault_core.totalValue()
OLD_AAVE_STRAT = AAVE_STRAT
NEW_AAVE_STRAT = "0xaB5E7B701B605f74AaC1b749Fd50715f0DEd7Bc5"
COMPOUND_STRAT = '0xD5433168Ed0B1F7714819646606DB509D9d8EC1f'
AAVE_INCENTIVES_CONTROLLER = "0xd784927Ff2f95ba542BfC824c8a8a98F3495f6b5"
ADAI = "0x028171bca77440897b824ca71d1c56cac55b68a3"
new_aave_strat = Contract(NEW_AAVE_STRAT)
incentives = Contract(AAVE_INCENTIVES_CONTROLLER)
stkAave = Contract('0x4da27a545c0c5b758a6ba100e3a049001de870f5')
aave = Contract('0x7Fc66500c84A76Ad7e9c93437bFc5Ac33E2DDaE9')
adai = Contract(ADAI)
brownie.network.web3.provider.make_request('hardhat_impersonateAccount', ['0x830622bdd79cc677ee6594e20bbda5b26568b781'])
brownie.network.web3.provider.make_request("evm_snapshot", [])

new_aave_strat.checkBalance(DAI) / 1e18
vault_admin.reallocate(COMPOUND_STRAT, NEW_AAVE_STRAT, [DAI], [int(1000000*1e18)], {'from': GOVERNOR})
vault_admin.reallocate( NEW_AAVE_STRAT, COMPOUND_STRAT, [DAI], [int(1000000*1e18)], {'from': GOVERNOR})
vault_admin.reallocate(COMPOUND_STRAT, NEW_AAVE_STRAT, [DAI], [int(2000000*1e18)], {'from': GOVERNOR})
new_aave_strat.checkBalance(DAI) / 1e18

for i in range(1,30):
  print("---- DAY %d" %i)
  #brownie.network.web3.provider.make_request('evm_mine', [1627490205 + 86400 * i])
  brownie.network.web3.provider.make_request('evm_increaseTime', [86400])
  print("USDT %d" % (usdt.balanceOf(vault_core) / 1e6))
  print("STK %d" % (stkAave.balanceOf(NEW_AAVE_STRAT) / 1e18))
  print("DAI %d" % (new_aave_strat.checkBalance(DAI) / 1e18))
  print("Cooldown %d" % stkAave.stakersCooldowns(NEW_AAVE_STRAT))
  print("RB %d" % incentives.getRewardsBalance([ADAI], new_aave_strat))
  vault_admin.harvest(NEW_AAVE_STRAT, {'from':GOVERNOR})

#vault_admin.harvest(NEW_AAVE_STRAT, {'from':GOVERNOR})
```

**Comment by DanielVF** [comment]
@franckc, @tomlinton, @mikeshultz. This code is ready for review again, small changes to the contract and to the deploy script. I was able to run success path fork testing.

File: contracts/contracts/strategies/AaveStrategy.sol [hunk 1]
```diff
@@ -11,9 +11,47 @@ import {
     InitializableAbstractStrategy
```
Comment (tomlinton) [inline contracts/contracts/strategies/AaveStrategy.sol:11]: Is there any merit in leaving this strategy untouched as a V1 integration and creating a new file AaveStrategyV2 or similar?

Comment (DanielVF) [reply contracts/contracts/strategies/AaveStrategy.sol:11]: Only if we thought it likely that we would use AAVE v1 again. But we can aways resurrect this from the Git history if we really need it.

File: contracts/deploy/020_aave_v2.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    );
+    // 3. Init the proxy to point at the implementation
+    await withConfirmation(
+      cAaveStrategyProxy
+        .connect(sDeployer)
+        ["initialize(address,address,bytes)"](
+          dAaveStrategyImpl.address,
+          deployerAddr,
+          []
+        )
```
Comment (franckc) [inline contracts/deploy/020_aave_v2.js:51]: missing the getTxOpts() which is super helpful to speed up tx by bumping up gas price
same comment for the other calls to withConfirmation that don't include it in this file.

**Review by franckc — COMMENTED** [review]
Reviewed the deploy script - minor comment inline otherwise lgtm.
I haven't reviewed the contract piece yet.

File: contracts/deploy/020_aave_v2.js (hunk 2 above)
Comment (DanielVF) [reply contracts/deploy/020_aave_v2.js:52]: Updated. Thanks!

File: contracts/contracts/strategies/AaveStrategy.sol [hunk 3]
```diff
@@ -170,19 +213,71 @@ contract AaveStrategy is InitializableAbstractStrategy {
...
+
+        // Check staked AAVE cooldown timer
+        uint256 cooldown = stkAave.stakersCooldowns(address(this));
+        uint256 windowStart = cooldown + stkAave.COOLDOWN_SECONDS();
+        uint256 windowEnd = windowStart + stkAave.UNSTAKE_WINDOW();
+        uint256 currentTimestamp = now;
+
+        // If inside the unlock window, then we can redeem stkAave
+        // for AAVE and send it to the vault.
+        if (currentTimestamp > windowStart && currentTimestamp < windowEnd) {
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/strategies/AaveStrategy.sol]: Thanks. Corrected.

File: contracts/contracts/strategies/AaveStrategy.sol [hunk 4]
```diff
@@ -170,19 +213,71 @@ contract AaveStrategy is InitializableAbstractStrategy {
...
+    function collectRewardToken() external onlyVault nonReentrant {
+        if (address(stkAave) == address(0)) {
+            return;
+        }
+
+        // Check staked AAVE cooldown timer
+        uint256 cooldown = stkAave.stakersCooldowns(address(this));
+        uint256 windowStart = cooldown + stkAave.COOLDOWN_SECONDS();
+        uint256 windowEnd = windowStart + stkAave.UNSTAKE_WINDOW();
+        uint256 currentTimestamp = now;
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/strategies/AaveStrategy.sol]: I prototyped this in python, then converted it to solidity, so I had this as a var since it was an external RPC call at the time.

Turns out "now" is only 2 gas, so I've just replaced currentTimestamp with it through the code.

Thanks!

**Comment by DanielVF** [comment]
@tomlinton, could you take a look at the contract side of this when you have a chance?

File: contracts/contracts/strategies/AaveStrategy.sol [hunk 5]
```diff
@@ -170,19 +213,71 @@ contract AaveStrategy is InitializableAbstractStrategy {
...
+            uint256 stkAaveBalance = stkAave.balanceOf(address(this));
+            if (stkAaveBalance > rewardLiquidationThreshold) {
+                stkAave.redeem(address(this), stkAaveBalance);
+            }
+            // Transfer AAVE to vaultAddress
+            uint256 aaveBalance = IERC20(rewardTokenAddress).balanceOf(
+                address(this)
+            );
+            if (aaveBalance > 0) {
+                IERC20(rewardTokenAddress).safeTransfer(
```
Comment (tomlinton) [inline contracts/contracts/strategies/AaveStrategy.sol:241]: The Uniswap path that we use to liquidate these is hardcoded as reward token -> weth -> usdt. Did you check if that is suitable here (and for Compound for that matter). My guess is that it is, but I've always had a bit of an uneasy feeling about it.

**Comment by tomlinton** [comment]
Not sure if I previously suggested this, but any thoughts on maintaining the v1 strategy as is (in the codebase) and adding a new file for the v2 strategy stuff? Might be more pain than it is worth.

File: contracts/contracts/strategies/AaveStrategy.sol [hunk 6]
```diff
@@ -170,19 +213,71 @@ contract AaveStrategy is InitializableAbstractStrategy {
...
+                    vaultAddress,
+                    aaveBalance
+                );
+            }
+        }
+
+        // If we were past the start of the window,
```
Comment (tomlinton) [inline contracts/contracts/strategies/AaveStrategy.sol:248]: This comment is a bit confusing. I think it needs to be updated to mention claiming of rewards. "If we have passed the start of the window, claim any outstanding rewards. If the cooldown counter is not running then start it."

**Comment by tomlinton** [comment]
I think the reason I suggested the v1 vs v2 thing is that when I review the changes here I've got the blinders on with regard to how the rest of the strategy works because the stuff we originally had doesn't show up in the diff, e.g. checkBalance. Fortunately it's all reasonably simple.

File: contracts/contracts/strategies/AaveStrategy.sol (hunk 5 above)
Comment (DanielVF) [reply contracts/contracts/strategies/AaveStrategy.sol:241]: I checked and that's currently the best route on both AAVE and COMP. Looks like were are good.

**Comment by DanielVF** [comment]
> Not sure if I previously suggested this, but any thoughts on maintaining the v1 strategy as is (in the codebase) and adding a new file for the v2 strategy stuff? Might be more pain than it is worth.

Yeah, maybe talked in Discord. I'd rather keep the codebase simple, and we can all ways get the old strategy back from git in the unlikely event that we need it.

File: contracts/contracts/strategies/AaveStrategy.sol [hunk 7]
```diff
@@ -70,7 +108,12 @@ contract AaveStrategy is InitializableAbstractStrategy {
 
         IAaveAToken aToken = _getATokenFor(_asset);
         emit Withdrawal(_asset, address(aToken), _amount);
-        aToken.redeem(_amount);
+        uint256 actual = _getLendingPool().withdraw(
+            _asset,
+            _amount,
+            address(this)
+        );
+        require(actual >= _amount, "Did not withdraw enough");
```
[the remark this answers is not in the document]
Comment (DanielVF) [reply contracts/contracts/strategies/AaveStrategy.sol:116]: I guess this was just a cheap sanity check.  Do you think it would be better to do `actual == amount`?

File: contracts/contracts/strategies/AaveStrategy.sol (hunk 6 above)
Comment (DanielVF) [reply contracts/contracts/strategies/AaveStrategy.sol]: Updated the comment.

File: contracts/contracts/strategies/AaveStrategy.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikeshultz) [reply contracts/contracts/strategies/AaveStrategy.sol:116]: IMO, I think that would be more appropriate since `actual` should never be greater than `_amount`.  I was also suggesting the revert message wasn't accurate.

**Review by tomlinton — APPROVED** [review]
👍

**Comment by DanielVF** [comment]
#### Requirements

This PR moves from AAVE v1 to AAVE v2, and adds support for the AAVE v2 rewards.

Sheesh StkAAVE is a pain.

### Deployment Considerations

AAVE v2 rewards collection requires that the collectRewardToken() method on the strategy be called by the vault on the correct times to hit the unstaking window. We may need a keeper contract to do this.


#### Internal State

No internal state is modified during operation. Contract storage slots are configuration. All dynamic state is looked up from external sources.

💡 If we were using a more recent solidity version, we could `immutable` all these configuration addresses into the contract, and save gas on everything the contract does.

#### Attack

AAVE very difficult to misuse.

Funds are only ever transferred out to the vault, or to the address specified by the vault.

Balances are a simple 1-1 mapping of aToken balances.


#### Logic

💡 We aren't handling collecting the small amounts of StkAAVE interest that could be generated from the week worth of StkAAVE we'll build up between converting it to AAVE. If AAVE reintroduces rewards mining, we can upgrade to support it.


#### Tests

  - [ ] Each logical branch has a test
  - [ ] Each require() has a test
  - [ ] Edge conditions are tested

#### Flavor

_Could reorder methods for more standardization. Will do later though, if we do._

#### Overflow

- [x] Never use "+" or "-", always use safe math
- [x] Check that all for loops use uint256

#### Black magic

- [x] Does not contain `selfdestruct`

- [x] Does not use `delegatecall` outside of proxying

- [x] (If an implementation contract were to call delegatecall under attacker control, it could call selfdestruct the implementation contract, leading to calls through the proxy silently succeeding, even though they were failing.)

- [x] Address.isContract should be treated as if could return anything at any time, because that's reality.


#### Authentication

- [x] Never use tx.origin
- [x] Check that every external/public function should actually be external
- [x] Check that every external/public function has the correct authentication

#### Cryptographic code

_No crypto code used._

#### Gas problems

- [x] Contracts with for loops must have either:
    - [ ] A way to remove items
    - [x] Can be upgraded to get unstuck
    - [x] Size can only controlled by admins
- [x] Contracts with for loops must not allow end users to add unlimited items to a loop that is used by others or admins.

#### External calls

- [x] Contract addresses passed in are validated
- [x] Unsafe external calls. _AAVE and trusted stables only for external calls_
- [x] Reentrancy guards on all state changing functions
- [x] Malicious behaviors. _If AAVE is malicious, problems ensue._
- [x] Could fail from stack depth problems (low level calls much require success)
- [x] No slippage attacks (we need to validate expected tokens received)
- [x] Oracles? _none_
- [x] Oracles, one of:
  - [x] No oracles
  - [ ] Oracles can't be bent
  - [ ] If oracle can be bent, it won't hurt us.
- [x] Don't call balanceOf for external contracts to determine what they will do, when they instead use internal accounting?

#### Ethereum

- [x] Contract does not send or receive Ethereum.
- [x] Contract has no payable methods.
- [x] Contract does not use msg.value
